### PR TITLE
Add ItemStateUpdatedEvent and enable group channel-links

### PR DESCRIPTION
**Pull request by J-N-K**
There has been some discussion in the past about the `ItemStateEvent` which in some cases may not carry the unit of a value even if the item has a unit (#2956).

This PR adds an `ItemStateUpdatedEvent` which is fired by the item itself (similar to the `ItemStateChangedEvent`. Since the item knows about its dimension and unit, the new event can carry this information even if the unit is derived from a unit provider or the state description. This also clearly separates the events when the item reports something and when the item is supposed to do something:

- `ItemStateEvent` -> the item shall update to the contained state
- `ItemCommandEvent`-> the item receives this command
- `ItemStateUpdatedEvent` -> the item updated the state
- `ItemStateChangedEvent` -> the item changed the state

Since this is rather big change in functionality I would suggest to consider this for OH4 because time is running out for extensive testing (even if unit/integration test coverage is quite good).

As a side-effect group items now also report their state which makes it possible to send aggregated states (AVG/SUM/MIN/MAX/...) to the outside world by using the FOLLOW profile.

### CODE REVIEW

**Comment by spacemanspiff2007** [comment]
Awesome! I really think the differentiation between report and todo makes things more clear for the user.
Could you also an example how this would look as an SSE event?

File: bundles/org.openhab.core/src/main/java/org/openhab/core/items/events/GroupStateUpdatedEvent.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.core.types.State;
+
+/**
+ * {@link GroupStateUpdatedEvent}s can be used to deliver group item state updates through the openHAB event bus.
+ * In contrast to the {@link GroupItemStateChangedEvent} it is always sent.
+ * State events must be created with the {@link ItemEventFactory}.
+ *
+ * @author Ja n N. Klug - Initial contribution
```
Comment (jlaur) [inline bundles/org.openhab.core/src/main/java/org/openhab/core/items/events/GroupStateUpdatedEvent.java:24]: Minor 😄
```suggestion
 * @author Jan N. Klug - Initial contribution
```

File: bundles/org.openhab.core/src/test/java/org/openhab/core/library/items/NumberItemTest.java [hunk 2]
```diff
@@ -40,6 +49,8 @@
 import org.openhab.core.types.UnDefType;
 import org.openhab.core.types.util.UnitUtils;
 
+import tech.units.indriya.unit.Units;
```
Comment (ssalonen) [inline bundles/org.openhab.core/src/test/java/org/openhab/core/library/items/NumberItemTest.java:52]: Is this correct import?

**Comment by J-N-K** [comment]
@spacemanspiff2007 Looks like the `ItemStateEvent`, it uses the same payload. It has a different topic (`/stateupdated`, in line with `/statechanged`) and always contains the new state of the item. 

Regarding the timestamp: IMO it's still better to set the timestamp on the state, as that allows handling historic and future states with the same logic we have now. I wouldn't veto an addition to the events, but that should be a change for all events, not just some, so it's out of scope here.

**Comment by J-N-K** [comment]
@openhab/core-maintainers Can we agree that this is useful and merge it? It makes the implementation of a solution for #3282 much easier because we have an event that carries the correct unit for state updates (which is needed e.g. for persistence).

**Comment by spacemanspiff2007** [comment]
Additionally it's somewhat API breaking for RestAPI consumers because 
- The new event should be used
- The legacy code doing the event state to item state magic can be completely removed 
  (e.g. `ItemStateEvent(1)` to `Switch` results in `ON` but to Number results in `1`).

It's a huge simplification and makes dealing with the RestAPI events much more pleasant.

**Review by kaikreuzer — APPROVED** [review]
Thank you!

**Comment by lolodomo** [comment]
I have difficulty to understand the impact of this change on the remote OH binding. 
Is it ok to ignite this new event and only rely on old events? 
Should I ignore ItemStateEvent and consider rather ItemStateUpdatedEvent when and only when the remote server is OH 4 ? 

https://github.com/openhab/openhab-addons/blob/main/bundles/org.openhab.binding.remoteopenhab/src/main/java/org/openhab/binding/remoteopenhab/internal/rest/RemoteopenhabRestClient.java#L416

**Comment by lolodomo** [comment]
Is it just an additional event that can be ignored or is the behavior with the previous events impacted?

**Comment by J-N-K** [comment]
There is a benefit in using it, but it does no harm ignoring it. All events are still fired like before.

One of the limitations of `ItemStateEvent` is that it does not contain information about the real state the item has. E.g. a `ON` event to a `DimmerItem` changes the internal state to `100%` because internally the representation of the state is `PercentType`. The `ItemStateChangedEvent` however does report `100%`. The `ItemStateUpdatedEvent` also contains the new state of the item in the same way it is stored internally. The same applies to `DecimalTypes` sent to UoM-`Number` items, `Color` items and roller shutters.

File: bundles/org.openhab.core/src/main/java/org/openhab/core/items/GenericItem.java [hunk 3]
```diff
@@ -231,14 +231,23 @@ protected final void applyState(State state) {
...
+            eventPublisher1.post(ItemEventFactory.createStateUpdatedEvent(this.name, newState, null));
+        }
+    }
+
     private void sendStateChangedEvent(State newState, State oldState) {
-        if (eventPublisher != null) {
-            eventPublisher.post(ItemEventFactory.createStateChangedEvent(this.name, newState, oldState));
+        EventPublisher eventPublisher1 = this.eventPublisher;
+        if (eventPublisher1 != null) {
+            eventPublisher1.post(ItemEventFactory.createStateChangedEvent(this.name, newState, oldState));
```
[the remark this answers is not in the document]
Comment (J-N-K) [reply bundles/org.openhab.core/src/main/java/org/openhab/core/items/GenericItem.java:250]: The `eventPublisher` is `@Nullable`. There is a slight chance that it is set to `null` by another thread between the null-check and the usage. Therefore taking a copy ensure we don't get a NPE when using it: even if the field is set to `null`, we still have the non-null reference.

[the remark this answers is not in the document]
Comment (jimtng) [reply bundles/org.openhab.core/src/main/java/org/openhab/core/items/GenericItem.java:250]: Thank you! TIL!

**Comment by J-N-K** [comment]
What exactly is broken? The `ItemStateEvent` was replaced by the `ItemStateUpdatedEvent` because that is what contains the new item's state, but I don't see why that would break something?

**Comment by florian-h05** [comment]
For file-based scripts you can create rules using `rules.JSRule`, the callback functions of these are provided the event data that triggered the rule run. The event is not passed 1:1 to the rule, because handling Java types in JS is confusing. Instead, based on the classname of the event, properties are added to a pure JS object. 

I just need to adjust the classnames there.

> The ItemStateEvent was replaced by the ItemStateUpdatedEvent

I‘ve already figured out that, thanks.

Did anything change for group events? On first look, it seems that `GroupItemStateEvent` was replaced by `ItemUpdatedEvent`.

**Comment by J-N-K** [comment]
`GroupItemStateEvent` that ever existed, even though it was in the JavaDoc. This PR introduced the updated event for groups (which was not reported before).

**Comment by jimtng** [comment]
I'm wondering why ItemStateUpdatedEvent extends `ItemEvent` and not `ItemStateEvent`. If it extended `ItemStateEvent` I think the jruby library would've been unaffected. But this is not a big problem and I've got a fix pending for it. 


The thing that puzzled me though is that previously (e.g. in openhab 3.4), I can send BusEvent.postUpdate("GroupItem", ON) and this will trigger an Item updated event for "GroupItem".

This stopped working in 4.0 recently. I don't know when exactly, but I know that it wasn't *this* PR that changed that behaviour. I looked around in the core code but even in 3.4 I can't actually find where this would have triggered, so the fact that it triggered in 3.4 is a mystery to me, but it did. We have a test to verify this behaviour since late last year and it has always passed until recently.

**Comment by lolodomo** [comment]
No impact on DSL rules?

**Comment by J-N-K** [comment]
I don't think so. They just pass the event payload and that is the same (except that you really get the item state instead of something that may or may not be the item state).

**Comment by creativeprojects** [comment]
Is there any plan to deprecate the `ItemStateEvent` in the future?
I understand why the new one is better, but then I don't think the original event serves any purpose any more.

**Comment by J-N-K** [comment]
It still has an important purpose: it is issued to initiate an item state update. So the `ItemStateEvent` results in an update of the item, which is then communicated as `ItemStateUpdatedEvent` (and potentially an `ItemStateChangedEvent`).

**Comment by creativeprojects** [comment]
OK.
So if I understand well, it could mean that if a channel linked to the item state is down, we might only get `ItemStateEvent` but not a `ItemStateEvent` in that case?

**Comment by spacemanspiff2007** [comment]
No.
`ItemStateEvent`: Event that tells the item to update its state
`ItemStateUpdatedEvent`: Event that tells that an item has updated it's state

Do you see the subtle but important difference.

Let's take a DimmerItem:
`ItemStateEvent`: `ON`  (Item should turn on)
`ItemUpdatedEvent`: `100%` (Dimmer that is on has 100%)

**Comment by creativeprojects** [comment]
Right, I understand the difference, thank you for the explanation 👍🏻 

I think we could update the documentation with this information: https://www.openhab.org/docs/developer/utils/events.html

**Pull request merged**